### ISSUE TITLE
Support MapBox

### ISSUE THREAD

**Issue by shesio** [p-lr/MapComposeMP#3]
Please add support for MapBox maps in MapComposeMP. Thank you!

**Comment by p-lr** [comment]
Yes, I'm aware that would be nice.

**Comment by jershell** [comment]
I'm taking this issue. If anyone else is working on it too, please let me know so we can coordinate and avoid duplicate work.

**Comment by brezinajn** [comment]
I've made some initial progress in Mapbox integration. If you want to compare notes/colab, I'm all for it.

**Comment by jershell** [comment]
> I have rough but almost complete impl
Did you implement the full specification?
The most difficult trouble it`s expressions and filters

https://maplibre.org/maplibre-style-spec/expressions/
https://maplibre.org/maplibre-style-spec/layers/#filter

https://github.com/mapbox/tilejson-spec/blob/master/2.2.0/README.md

**Comment by jershell** [comment]
🚧 Problem Overview

I've encountered a structural mismatch between how tile zooming is handled in MapCompose versus MapLibre. We should decide how we want to handle this going forward.

### 📽 "Stupid" rasterization from MVT to PNG
[▶️ Watch the video](https://raw.githubusercontent.com/jershell/MapComposeMP/feature/mvt-support/doc/wip/trouble_direct_raster.mp4)

---

### ❌ Why treating vector tiles as if they were raster is problematic

When vector tiles are processed as if they were already rasterized (e.g., simply fetched and painted onto the screen like static images), this goes against the core principles of vector rendering and introduces several serious issues:

1. **Excessive number of tile requests**
2. **Loss of vector rendering benefits** (scalability, crispness on high-DPI screens)
3. **Incorrect scaling of symbols and text**
4. **Layer filtering logic becomes ineffective**
5. **Client-side performance degradation** (due to increased draw calls and memory usage)
6. **Broken interactivity and hit-testing** (you can’t query vector features anymore)

---

### ✅ Correct approach

* Only one tile should be rendered at the lowest zoom level
* All logic related to scaling, filtering, and styling should be handled on the client side
* The client should decide which features to render, how to scale text and lines, and how to apply style rules dynamically

---

### 🎯 Key point

`MapCompose` does not stretch a single tile to fill the entire screen. Instead, it always divides the world into a fixed `fullSize × fullSize` pixel grid and tries to cover the viewport using as many tiles as necessary.

If the viewport is larger than the rendered world at a certain zoom level, it starts wrapping or requesting “extra” tiles to fill the space.

---

### Let's we treat `MapLibre` as the reference standard

`MapCompose` (raster) and `MapLibre` (vector) interpret the world at minimum zoom very differently:

* In **MapLibre**, zoom level 0 always renders a **single tile (0/0/0)** that covers the entire world.
* In **MapCompose**, as many tiles are rendered as needed to fill the viewport, even at the lowest zoom.

To match MapLibre's behavior precisely, we either need to:

* Patch `MapCompose`, or
* Apply a workaround in the tile provider (e.g., always returning only tile `0/0/0` at zoom 0).

---

### What MapLibre does

1. At `zoom = 0`, the logical world is covered by **1 tile of 256×256 pixels**
   This is based on the Web Mercator projection (EPSG:3857), where the world is projected into a fixed number of tiles.

   * At `zoom = 0`: 1 tile (256×256) covers the entire world
   * At `zoom = 1`: 2×2 tiles
   * And so on: tile matrix = `2^zoom × 2^zoom`

👉 This tile matrix model is a foundational concept used in OpenMapTiles, Mapbox, MapLibre, Google Maps, and similar systems.

---

### What MapCompose currently does

`MapCompose` calculates tile layout based on `scale` and `fullSize`, not based on the standardized tile matrix.

As a result, there’s **no fixed tile matrix** for a given zoom level. For example, at the lowest zoom level (`level = 0`), it may load 4, 6, or 9 tiles depending on the viewport size — which violates the MVT (Mapbox Vector Tile) specification.

---

## 🔧 What needs to be changed in `MapCompose`

### 1. Fix `scaleForLevel` to follow the tile matrix model:

**Current:**

```kotlin
scaleForLevel[level] = 1.0 / 2.0.pow((levelCount - level - 1))
```

**Should be:**

```kotlin
scaleForLevel[level] = 1.0 / 2.0.pow(level)
```

This change makes `level == zoom`, instead of `level = maxZoom - zoom`.

---

### 2. Update `getLevel(scale)` logic:

**Current:**

```kotlin
val partialLevel = levelCount - 1 - magnifyingFactor + ln(scale) / ln(2.0)
```

**Should be:**

```kotlin
val partialLevel = log2(1.0 / scale)
return partialLevel.roundToInt().coerceIn(0, levelCount - 1)
```

---

### 3. Bind `fullSize` to the tile matrix size:

Instead of using a fixed pixel size (e.g., `16384 x 16384`), use:

```kotlin
fullSize = tileSize * 2^maxZoom
```

This ensures the world is correctly scaled across all zoom levels. If `fullSize` is inconsistent (e.g., 18000 pixels), the matrix alignment breaks.

---

### 4. Add support for world-wrapping

MapLibre **wraps the world horizontally**, allowing infinite panning.

MapCompose **does not**. To replicate expected behavior:

* Enable horizontal wrap-around, or
* Lock scroll boundaries to the world bounds.

---
(https://docs.mapbox.com/help/glossary/zoom-level/) or [[EPSG:3857](https://epsg.io/3857)](https://epsg.io/3857)

**Comment by p-lr** [comment]
I have to point out that MapCompose can be configured to be used using standard WMTS tile source. In the OSM demo, you can see that the map size si computed like so:

```
fun mapSizeAtLevel(wmtsLevel: Int, tileSize: Int): Int {
    return tileSize * 2.0.pow(wmtsLevel).toInt()
}
```
In this case, there is a perfect match between the level in the library and the level you're talking about (let's call it the `wmts level`). It's just that MapCompose can represent other maps which aren't necessarily world maps. I have an app which uses MapCompose to display small maps (of some forests, islands, etc). Once you know the extent of the captured area, you can make to correspondance between with the wmts level if needed. It really is an implementation detail of the `TileStreamProvider`.

As for your 4th point, you're absolutely right. MapCompose isn't designed for infinite scrolling. That could be something to work on.

But if we have a feature complete renderer which bascically does the pbf -> png conversion, that would be huge already.

**Comment by jershell** [comment]
> complete renderer which bascically does the pbf -> png conversion


Great! I’ll focus solely on rasterization (PBF → PNG) for now. Once the MR is ready, we can continue with integrating the raster output into MapCompose.

**Comment by DestBro** [comment]
> This is the real challenge indeed. With this additional step, we can finally have a complete Kotlin library to display vector tiles.

Is this done ? Any samples on how to use vector tiles map with MapCompose ?

**Comment by p-lr** [comment]
> Is this done ? Any samples on how to use vector tiles map with MapCompose ?

No. @jershell didn't give any update for quite a while now. He either stopped working on it or forked the project :/

**Comment by jershell** [comment]
> > This is the real challenge indeed. With this additional step, we can finally have a complete Kotlin library to display vector tiles.
> 
> Is this done ? Any samples on how to use vector tiles map with MapCompose ?

In this branch (https://github.com/p-lr/MapComposeMP/tree/feature/mvt-support) we are working on MVT tile support, and in this PR (https://github.com/p-lr/MapComposeMP/pull/29) you can find screenshots of the current progress in the description.

> > Is this done ? Any samples on how to use vector tiles map with MapCompose ?
> 
> No. [@jershell](https://github.com/jershell) didn't give any update for quite a while now. He either stopped working on it or forked the project :/

Regarding further progress, we’ve run into some difficulties with the 3D layer. It’s necessary to implement rendering of 3D objects at the engine level and use the current rendering result as a texture. This part raises the most questions, and I haven’t started exploring it yet. The previous work I did during my vacation, but once it ended, work life started “attacking” me and kept me extremely busy. Sadly, I’m not sure when I’ll be able to continue my research.

**Comment by rkreienbuehl** [comment]
> > Regarding further progress, we’ve run into some difficulties with the 3D layer. It’s necessary to implement rendering of 3D objects at the engine level and use the current rendering result as a texture. This part raises the most questions, and I haven’t started exploring it yet. The previous work I did during my vacation, but once it ended, work life started “attacking” me and kept me extremely busy. Sadly, I’m not sure when I’ll be able to continue my research.
> 
> Interesting. Do you think it would be worth to target a first milestone without support for 3D layer? If so, as you might not have time to wrap all what's been done until now, I might try to do it.

@jershell @p-lr It would be great to see a release without 3d support. I am also willing to help if there is something I can do?

**Comment by rkreienbuehl** [comment]
@p-lr @jershell 
I now have a rough overview of the implementation. Maybe it is good to shortly discuss, how the vectortile integration should look in the end, especially because @jershell did an excellent job so far, and I don't want to decide something, that doesn't match with his thoughts.
At the moment the `maplibre-rasterizer` is like an addition to MapCompose that can be used to write a `TileStreamProvider`, which integrates an own httpClient.

My suggestion would be to create a new `VectorTileStreamProvider` interface, which should include the style as well as a `getTileStream` function. This provider could be added to the map with an additional function `addLayer`, which takes a `VectorTileStreamProvider`.
Conceptually, this would be the same as raster layers. The user would be responsible for loading and caching, but the drawing would be handled by MapCompose.

This would bring the following changes:
- move `maplibre-rasterizer` into the `library`
- creating a new Interface `VectorTileStreamProvider ` and extension function `addLayer`
- remove `MapLibreRasterizer` and use parts of the code inside `addLayer` function (or change `MapLibreRasterizer` so that loading and caching would be done outside and use it in `addLayer`)
- maybe add/expose some helper functions (not sure if needed)

This is the plan I put together in my mind, so maybe it's not complete. What do you think about it this approach?

**Comment by zoff99** [comment]
is this now supported? can i test it?

**Comment by rkreienbuehl** [comment]
@zoff99 The task is still with me. Unfortunately my time was very limited in the end of 2025, so I couldn't finish integrating the work already done by @jershell. Because I could also need it in the near future, it went up on my todo-list and I right today I started with it. I can't say any estimated time by now, but will communicate here on any news.

**Comment by zoff99** [comment]
thanks for your efforts. happy new year!

**Comment by zoff99** [comment]
any news on this topic?

**Comment by rkreienbuehl** [comment]
@zoff99 I finally had enough time to work properly on this. The integration of @jershell's work into the library is more or less done, but there are some issues to fix before this is really ready:

- demo with maptiler street_v2 style is implemented
- demo works well on desktop
- on iOS vector layer is really slow -> performance needs to be optimized
- on Android a vector layer crashes the app -> need to investigate this in depth
- collision detection is not yet integrated

Also, not related to integrating the work of @jershell into the library, the rendering itself has to be pursued further in order to support any vector source. For example I noticed, that some filters are missing that are needed.

@p-lr I will create a draft PR in the next days, so you can make a first review.

Overall it will take some time to finish this properly.

**Comment by p-lr** [comment]
@rkreienbuehl That's great news. Vector tiles is clearly the next milestone for this library. 
There's another issue, and you probably saw it too: there's a variety of styles format. Mapbox develop their own extensions and features, while MapLibre strives to define standards. The questions: "Which style format does MapCompose support?", or "Which vector tiles source are supported?" should have clear answers. I'm curious to know what you think about that.

Looking forward to see your work (and to seriously start working on it :)

**Comment by rkreienbuehl** [comment]
@p-lr I made a draft PR for a first look on the work.
I will look into the differences of MapBox and Maplibre and share my thoughts.

**Comment by zoff99** [comment]
could you please add a screenshot to show an example map with mapbox?
